types/wlr_wl_shell.c

```diff
@@ -1,61 +1,264 @@
-		*resource, struct wl_resource *seat, uint32_t serial) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface move");
+static void shell_surface_move(struct wl_client *client,
+		struct wl_resource *resource, struct wl_resource *seat_resource,
+		uint32_t serial) {
+	wlr_log(L_DEBUG, "got shell surface move");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+	struct wlr_seat_handle *seat_handle =
+		wl_resource_get_user_data(seat_resource);
+
+	struct wlr_wl_shell_surface_move_event *event =
+		calloc(1, sizeof(struct wlr_wl_shell_surface_move_event));
+	if (event == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	event->client = client;
+	event->surface = surface;
+	event->seat_handle = seat_handle;
+	event->serial = serial;
+
+	wl_signal_emit(&surface->events.request_move, event);
+
+	free(event);
 }
 
 static void shell_surface_resize(struct wl_client *client,
-		struct wl_resource *resource, struct wl_resource *seat, uint32_t serial,
-		uint32_t edges) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface resize");
+		struct wl_resource *resource, struct wl_resource *seat_resource,
+		uint32_t serial, uint32_t edges) {
+	wlr_log(L_DEBUG, "got shell surface resize");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+	struct wlr_seat_handle *seat_handle =
+		wl_resource_get_user_data(seat_resource);
+
+	struct wlr_wl_shell_surface_resize_event *event =
+		calloc(1, sizeof(struct wlr_wl_shell_surface_resize_event));
+	if (event == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	event->client = client;
+	event->surface = surface;
+	event->seat_handle = seat_handle;
+	event->serial = serial;
+	event->edges = edges;
+
+	wl_signal_emit(&surface->events.request_resize, event);
+
+	free(event);
 }
 
 static void shell_surface_set_toplevel(struct wl_client *client,
 		struct wl_resource *resource) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface set_toplevel");
+	wlr_log(L_DEBUG, "got shell surface toplevel");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+
+	if (surface->role != WLR_WL_SHELL_SURFACE_ROLE_NONE) {
+		return;
+	}
+
+	surface->role = WLR_WL_SHELL_SURFACE_ROLE_TOPLEVEL;
+
+	wl_signal_emit(&surface->events.set_role, surface);
 }
 
 static void shell_surface_set_transient(struct wl_client *client,
-		struct wl_resource *resource, struct wl_resource *parent, int32_t x,
-		int32_t y, uint32_t flags) {
-	wlr_log(L_DEBUG, "TODO: implement shell surface set_transient");
+		struct wl_resource *resource, struct wl_resource *parent_resource,
+		int32_t x, int32_t y, uint32_t flags) {
+	wlr_log(L_DEBUG, "got shell surface transient");
+	struct wlr_wl_shell_surface *surface = wl_resource_get_user_data(resource);
+	struct wlr_wl_shell_surface *parent =
+		wl_resource_get_user_data(parent_resource);
+	// TODO: check if parent_resource == NULL?
+
+	if (surface->role != WLR_WL_SHELL_SURFACE_ROLE_NONE) {
```

Comment:
We could probably use a function that changes the state to handle changes between the states. This is allowed.

Reply:
What do you mean?

Reply:
We need a way to change from one state to any other state.